### ISSUE TITLE
Destroy not stopping the rendering or removing the WebGL context

### ISSUE THREAD

**Issue by Nombezie** [sirxemic/jquery.ripples#37]
Love this plugin but I have ran into an anomaly.

I know it's not best practice, but bare with me.

AJAX-Page.html
- Button that uses AJAX  to load AJAX-Content.html

AJAX-Content.html
- Call ripple function on element in THIS page.

Smash the AJAX-Page.html and you're going to get a bunch of instances of the ripple function running on top of each other.

Even if you change AJAX-Content.html to:

AJAX-Content.html
- Call Destroy ripple on same as below element.
- Call ripple function on element in this page.

You end up with multiple ripples functions running on the same element.

In other words, Destroy doesn't seem to work when used with AJAX loaded content.

**Comment by sirxemic** [comment]
It's kinda hard to tell without seeing the code if this is an error in the plugin or in your code. I'm suspecting the latter. Are you sure you are destroying the plugin properly before applying it to an element?

**Comment by Nombezie** [comment]
Here is a rough demo for you.

http://www.bryan-myers.com/demo/ripple/

**demo.html**
```
<html lang="en">
  <head>
    <title>Demo</title>
  </head>
  <body>
    <div id="loader"></div>
    <style>
      body, html { margin: 0px; padding: 0px; width: 100%; height: 100%; }
      #loader { background-color: #333; height: 100vh; position: absolute; z-index: 0; width: 100%; }
    </style>
    <script src="https://code.jquery.com/jquery-1.12.4.js"></script>
    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>
    <script>
      $(document).ready(function() {
        $('#loader').load('content.html');
      });
    </script>
  </body>
</html>
```
**content.html**
```
<div id="myRipple">
  <div id="btnReload">Reload In Container</div>
</div>

<style>
  #myRipple{ height: 100vh; background-repeat: no-repeat; background-position: center; background-size: cover; background-color: #333; }
  #btnReload { padding: 20px; color: white; background: green; cursor: pointer; display: inline-block; }
</style>
<script src="/js/jquery.ripples.js"></script>
<script>
  $("#btnReload").click(function(event) {
    $('#loader').load('content.html');
  });
  $('#myRipple').ripples('destory');
  $('#myRipple').ripples({
    resolution: 512,
    dropRadius: 10,
    perturbance: 0.04
  });
</script>
```

**Comment by Nombezie** [comment]
As a side note, in Chrome, if you bring up the developer tools, you can see that after a few clicks of the button you will get this error:

WARNING: Too many active WebGL contexts. Oldest context will be lost.

**Comment by sirxemic** [comment]
First of all, there is a typo: `destory` instead of `destroy` ;)

Second of all, (just a guess) it could be that because you're loading `/js/jquery.ripples.js` multiple times you're getting unexpected behaviour.

**Comment by Nombezie** [comment]
I will update the demo.

**Comment by Nombezie** [comment]
I have moved the script include to the index.html page.
I have fixed the typo in the destroy call.

**index.html**

```
<html lang="en">
  <head>
    <title>Demo</title>
  </head>
  <body>
    <div id="loader"></div>
    <style>
      body, html { margin: 0px; padding: 0px; width: 100%; height: 100%; }
      #loader { background-color: #333; height: 100vh; position: absolute; z-index: 0; width: 100%; }
    </style>
    <script src="https://code.jquery.com/jquery-1.12.4.js"></script>
    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>
    <script src="/js/jquery.ripples.js"></script>
    <script>
      $(document).ready(function() {
        $('#loader').load('content.html');
      });
    </script>
  </body>
</html>
```

**content.html**
```
<div id="myRipple">
  <div id="btnReload">Reload</div>
</div>

<style>
  #myRipple{ height: 100vh; background-repeat: no-repeat; background-position: center; background-size: cover; background-color: #333; }
  #btnReload { padding: 20px; color: white; background: green; cursor: pointer; display: inline-block; }
</style>
<script>
  $("#btnReload").click(function(event) {
    $('#loader').load('content.html');
  });
  $('#myRipple').ripples('destroy');
  $('#myRipple').ripples({
    resolution: 512,
    dropRadius: 10,
    perturbance: 0.04
  });
</script>
```

**Comment by sirxemic** [comment]
Ok, maybe I misinterpreted the problem. What am I supposed to see?

**Comment by Nombezie** [comment]
Couple of other notes...  I have tried clearing the loader element a couple of ways as well.

$("#btnReload").click(function(event) {
    $('#loader').html("");
    $('#loader').load('content.html');
  });

And

$("#btnReload").click(function(event) {
    $('#loader').load('content.html');
  });
  myRipple = "";
  var myRipple = $('#myRipple').ripples({
    resolution: 512,
    dropRadius: 10,
    perturbance: 0.04
  });

**Comment by Nombezie** [comment]
Hit the reload button a bunch of times with or without the Chrome developer console open.  You will see that it is just instantiating multiple ripple objects.

**Comment by Nombezie** [comment]
I have also tried this with the random drop feature.  What you will end up with is multiple drops occurring at the same interval.  I can create a demo if you like.

**Comment by Nombezie** [comment]
If you get the warning, a refresh of the browser will reset the number of instances of the ripple object.

**Comment by Nombezie** [comment]
Another note, this is worsened by using a background image. I am guessing because each instance of the ripple object requires a lot more cpu with a background. I didn't include a background in the demo, but can easily add one.

**Comment by sirxemic** [comment]
Initially I thought you meant you'd actually *see* multiple canvases in the DOM, but you meant that there are multiple canvases being painted on, even after destroy, right?

I looked around a bit and I think I found the issue: by calling destroy the original element is restored to the original state and the canvas removed from the DOM. However, the 'process' which does the painting actually isn't stopped. It still updates the canvas, even though it's not part of the DOM anymore.

Good find! I think I'll fix it this weekend.

**Comment by Nombezie** [comment]
Damn dude! You are awesome! Web <3 my friend.

**Comment by sirxemic** [comment]
I released a version with a fix. Now it no longer continues rendering after it's destroyed, but do note it can take a while before the garbage collector actually removes the canvas.

**Comment by GorelovNA** [comment]
Hi, I have a completely identical problem - destroy method does not destroy the WebGL context behind itself. I use the latest version of the library. Can you fix it please? The library is awesome!
![Uploading Screenshot 2019-09-09 в 17.02.11.png…]()
